packages/loopring_v3/circuit/Gadgets/StorageGadgets.h

```diff
@@ -199,47 +213,54 @@ class StorageReaderGadget : public GadgetT {
   }
 };
 
-class NonceGadget : public GadgetT {
+// QUESTION(daniel): it seems this gadget is the same as ReadStorageGadget
+// except that it requires data to be 0.
+//
+// Maybe we should move this to `AccountGadgets.h' as this is on the account
+// level?
+class ReadNonceGadget : public GadgetT {
   const Constants &constants;
   const DualVariableGadget &storageID;
 
-  StorageReaderGadget storageReader;
+  ReadStorageGadget readStorage;
   IfThenRequireEqualGadget requireDataZero;
 
 public:
-  NonceGadget(ProtoboardT &pb, const Constants &_constants,
-              const StorageGadget &storage,
-              const DualVariableGadget &_storageID, const VariableT &verify,
-              const std::string &prefix)
-      : GadgetT(pb, prefix),
-
-        constants(_constants), storageID(_storageID),
-
-        storageReader(pb, constants, storage, storageID, verify,
-                      FMT(prefix, ".storageReader")),
-        requireDataZero(pb, verify, storageReader.getData(), constants._0,
-                        FMT(prefix, ".requireDataZero")) {}
+  ReadNonceGadget(ProtoboardT &pb, const Constants &_constants,
+              const StorageGadget &_currentStorage,
+              const DualVariableGadget &_storageID, const VariableT &_verify,
+              const std::string &_prefix)
+      : GadgetT(pb, _prefix),
+
+        constants(_constants), // QUESTION(daniel): can we not have the local
+                               // variables and
```

Comment:
These are stored here because `constants`/`storageID` is used in some of the other functions as well. But in the constructor we can of course use both.